Lib/fontTools/feaLib/builder.py

```diff
@@ -764,12 +764,25 @@ def buildGDEFGlyphClassDef_(self):
             classes = {g: c for (g, (c, _)) in self.glyphClassDefs_.items()}
         else:
             classes = {}
+            ignoreClassFlagsUsed = False
             for lookup in self.lookups_:
+                ignoreClassFlagsUsed = (
+                    ignoreClassFlagsUsed or (
+                        lookup.lookupflag & (
+                            otl.LOOKUP_FLAG_IGNORE_BASE_GLYPHS +
+                            otl.LOOKUP_FLAG_IGNORE_LIGATURES +
+                            otl.LOOKUP_FLAG_IGNORE_MARKS
+                        )
+                    )
+                )
                 classes.update(lookup.inferGlyphClasses())
+            markClassesUsed = 3 in classes.values()
             for markClass in self.parseTree.markClasses.values():
                 for markClassDef in markClass.definitions:
                     for glyph in markClassDef.glyphSet():
                         classes[glyph] = 3
+            if not markClassesUsed and not ignoreClassFlagsUsed:
+                classes.clear()
```

Comment:
why you clear after the loop? Before you enter the for loop above, you aready know whether markClassesUsed or ignoreClassFlagsUsed